app/grandchallenge/direct_messages/views.py

```diff
@@ -0,0 +1,304 @@
+                        default=Value(False),
+                        output_field=BooleanField(),
+                    )
+                ),
+            ),
+            "direct_messages__sender__user_profile",
+        )
+
+        return queryset
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data()
+
+        direct_message_form = DirectMessageForm(
+            sender=self.request.user, conversation=self.object
+        )
+        direct_message_form.helper.attrs.update(
+            {
+                "hx-post": reverse(
+                    "direct-messages:direct-message-create",
+                    kwargs={"pk": self.object.pk},
+                ),
+                "hx-target": "#conversation-detail-panel",
+            }
+        )
+
+        context.update(
+            {
+                "direct_message_form": direct_message_form,
+                "report_spam_form": DirectMessageReportSpamForm(),
+                "muted_users": {
+                    m.target: m.pk
+                    for m in Mute.objects.filter(
+                        source=self.request.user,
+                        target__in=self.object.participants.all(),
+                    )
+                },
+            }
+        )
+        return context
+
+
+class ConversationSelectDetail(
+    LoginRequiredMixin, ObjectPermissionRequiredMixin, DetailView
+):
+    permission_required = "direct_messages.view_conversation"
+    raise_exception = True
+    model = Conversation
+    template_name = "direct_messages/partials/conversation_select_detail.html"
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        return queryset.with_most_recent_message(user=self.request.user)
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data()
+
+        context.update(
+            {
+                "muted_users": {
+                    m.target: m.pk
+                    for m in Mute.objects.filter(
+                        source=self.request.user,
+                        target__in=self.object.participants.all(),
+                    )
+                },
+            }
+        )
+        return context
+
+
+class ConversationMarkRead(
+    LoginRequiredMixin, ObjectPermissionRequiredMixin, UpdateView
+):
+    permission_required = "direct_messages.mark_conversation_read"
+    raise_exception = True
+    model = Conversation
+    fields = ()
+
+    def form_valid(self, form):
+        DirectMessageUnreadBy.objects.filter(
+            direct_message__conversation=self.object,
+            unread_by=self.request.user,
+        ).delete()
+        return super().form_valid(form)
+
+
+class DirectMessageCreate(
+    LoginRequiredMixin, ObjectPermissionRequiredMixin, CreateView
+):
+    permission_required = "direct_messages.create_conversation_direct_message"
+    raise_exception = True
+    model = DirectMessage
+    form_class = DirectMessageForm
```

Comment:
Muted users can still create messages, and won't know that they have been muted by someone. I guess that's on purpose? Is that to protect the user who does the blocking? I feel like at least some sort of indication that the messages aren't making it through to the other user would be nice for the blocked user.